File: java/api-client/src/main/java/com/vmware/photon/controller/client/resource/VmApi.java
/*
 * Copyright 2015 VMware, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License.  You may obtain a copy of
 * the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed
 * under the License is distributed on an "AS IS" BASIS, without warranties or
 * conditions of any kind, EITHER EXPRESS OR IMPLIED.  See the License for the
 * specific language governing permissions and limitations under the License.
 */

package com.vmware.photon.controller.client.resource;

import com.vmware.photon.controller.api.model.ResourceList;
import com.vmware.photon.controller.api.model.Tag;
import com.vmware.photon.controller.api.model.Task;
import com.vmware.photon.controller.api.model.Vm;
import com.vmware.photon.controller.api.model.VmDiskOperation;
import com.vmware.photon.controller.api.model.VmMetadata;
import com.vmware.photon.controller.api.model.VmNetworks;
import com.vmware.photon.controller.client.RestClient;

import com.fasterxml.jackson.core.type.TypeReference;
import com.google.common.collect.ImmutableMap;
import com.google.common.util.concurrent.FutureCallback;
import org.apache.http.HttpResponse;
import org.apache.http.HttpStatus;

import javax.annotation.Nullable;

import java.io.IOException;

/**
 * Vm api.
 */
public class VmApi extends ApiBase {
  public VmApi(RestClient restClient) {
    super(restClient);
  }

  @Override
  public String getBasePath() {
    return "/vms";
  }

  /**
   * Get details about the specified vm.
   * @param vmId
   * @return Vm details
   * @throws java.io.IOException
   */
  public Vm getVm(String vmId) throws IOException {
    String path = String.format("%s/%s", getBasePath(), vmId);

    HttpResponse httpResponse = this.restClient.perform(RestClient.Method.GET, path, null);
    this.restClient.checkResponse(httpResponse, HttpStatus.SC_OK);

    return this.restClient.parseHttpResponse(
        httpResponse,
        new TypeReference<Vm>() {
        }
    );
  }

  /**
   * Get details about the specified vm.
   *
   * @param vmId
   * @param responseCallback
   * @throws IOException
   */
  public void getVmAsync(final String vmId, final FutureCallback<Vm>
      responseCallback) throws IOException {
    String path = String.format("%s/%s", getBasePath(), vmId);

    getObjectByPathAsync(path, responseCallback, new TypeReference<Vm>() {
    });
  }

  /**
   * Get tasks associated with the specified vm.
   * @param vmId
   * @return {@link ResourceList} of {@link Task}
   * @throws IOException
   */
  public ResourceList<Task> getTasksForVm(String vmId) throws IOException {
    String path = String.format("%s/%s/tasks", getBasePath(), vmId);

    ResourceList<Task> taskResourceList = new ResourceList<>();
    ResourceList<Task> resourceList = getTaskResourceList(path);
    taskResourceList.setItems(resourceList.getItems());
    while (resourceList.getNextPageLink() != null && !resourceList.getNextPageLink().isEmpty()) {
      resourceList = getTaskResourceList(resourceList.getNextPageLink());
      taskResourceList.getItems().addAll(resourceList.getItems());
    }

    return taskResourceList;
  }

  /**
   * Get all tasks at specified path.
   *
   * @param path
   * @return
   * @throws IOException
   */
  private ResourceList<Task> getTaskResourceList(String path) throws IOException {
    HttpResponse httpResponse = this.restClient.perform(RestClient.Method.GET, path, null);
    this.restClient.checkResponse(httpResponse, HttpStatus.SC_OK);
    return this.restClient.parseHttpResponse(
        httpResponse,
        new TypeReference<ResourceList<Task>>() {
        }
    );
  }

  /**
   * Get tasks associated with the specified vm.
   * @param vmId
   * @param responseCallback
   * @throws IOException
   */
  public void getTasksForVmAsync(final String vmId,
                                     final FutureCallback<ResourceList<Task>> responseCallback)
      throws IOException {
    String path = String.format("%s/%s/tasks", getBasePath(), vmId);

    ResourceList<Task> taskResourceList = new ResourceList<>();
    FutureCallback<ResourceList<Task>> callback = new FutureCallback<ResourceList<Task>>() {
      @Override
      public void onSuccess(@Nullable ResourceList<Task> result) {
        if (taskResourceList.getItems() == null) {
          taskResourceList.setItems(result.getItems());
        } else {
          taskResourceList.getItems().addAll(result.getItems());
        }
        if (result.getNextPageLink() != null && !result.getNextPageLink().isEmpty()) {
          try {
            getObjectByPathAsync(result.getNextPageLink(), this, new TypeReference<ResourceList<Task>>() {});
          } catch (IOException e) {
            e.printStackTrace();
          }
        } else {
          responseCallback.onSuccess(taskResourceList);
        }
      }

      @Override
      public void onFailure(Throwable t) {
        responseCallback.onFailure(t);
      }
    };

    getObjectByPathAsync(path, callback, new TypeReference<ResourceList<Task>>() {});
  }

  /**
   * Delete the specified vm.
   *
   * @param id - id of the vm to delete
   * @return Tracking {@link Task}
   * @throws IOException
   */
  public Task delete(String id) throws IOException {
    String path = String.format("%s/%s", getBasePath(), id);

    HttpResponse response = this.restClient.perform(RestClient.Method.DELETE, path, null);

    this.restClient.checkResponse(response, HttpStatus.SC_CREATED);
    return parseTaskFromHttpResponse(response);
  }

  /**
   * Delete the specified vm.
   * @param id
   * @param responseCallback
   * @throws IOException
   */
  public void deleteAsync(final String id, final FutureCallback<Task> responseCallback) throws IOException {
    deleteObjectAsync(id, responseCallback);
  }

  /**
   * Add the specified tag to the specified VM.
   * @param vmId - id of the vm
   * @param tag - {@link Tag}
   * @return Tracking {@link Task}
   * @throws IOException
   */
  public Task addTagToVm(String vmId, Tag tag) throws IOException {
    String path = String.format("%s/%s/tags", getBasePath(), vmId);

    HttpResponse httpResponse = this.restClient.perform(
        RestClient.Method.POST,
        path,
        serializeObjectAsJson(tag));
    this.restClient.checkResponse(httpResponse, HttpStatus.SC_CREATED);

    return parseTaskFromHttpResponse(httpResponse);
  }

  /**
   * Add the specified tag to the specified VM.
   * @param vmId
   * @param tag
   * @param responseCallback
   * @throws IOException
   */
  public void addTagToVmAsync(final String vmId, Tag tag, final FutureCallback<Task> responseCallback)
      throws IOException {
    String path = String.format("%s/%s/tags", getBasePath(), vmId);

    createObjectAsync(path, serializeObjectAsJson(tag), responseCallback);
  }

  /**
   * Perform a VM Start Operation on specified vm.
   * @param vmId - id of the vm
   * @return Tracking {@link Task}
   * @throws IOException
   */
  public Task performStartOperation(String vmId) throws IOException {
    String path = String.format("%s/%s/start", getBasePath(), vmId);

    HttpResponse httpResponse = this.restClient.perform(
        RestClient.Method.POST,
        path,
        null);
    this.restClient.checkResponse(httpResponse, HttpStatus.SC_CREATED);

    return parseTaskFromHttpResponse(httpResponse);
  }

  /**
   * Perform a VM Start Operation on specified vm.
   * @param vmId
   * @param responseCallback
   * @throws IOException
   */
  public void performStartOperationAsync(final String vmId,
                                         final FutureCallback<Task> responseCallback)
      throws IOException {
    String path = String.format("%s/%s/start", getBasePath(), vmId);

    createObjectAsync(path, null, responseCallback);
  }

  /**
   * Perform a VM Stop Operation on specified vm.
   * @param vmId - id of the vm
   * @return Tracking {@link Task}
   * @throws IOException
   */
  public Task performStopOperation(String vmId) throws IOException {
    String path = String.format("%s/%s/stop", getBasePath(), vmId);

    HttpResponse httpResponse = this.restClient.perform(
        RestClient.Method.POST,
        path,
        null);
    this.restClient.checkResponse(httpResponse, HttpStatus.SC_CREATED);

    return parseTaskFromHttpResponse(httpResponse);
  }

  /**
   * Perform a VM Stop Operation on specified vm.
   * @param vmId
   * @param responseCallback
   * @throws IOException
   */
  public void performStopOperationAsync(final String vmId,
                                         final FutureCallback<Task> responseCallback)
      throws IOException {
    String path = String.format("%s/%s/stop", getBasePath(), vmId);

    createObjectAsync(path, null, responseCallback);
  }

  /**
   * Perform a VM Restart Operation on specified vm.
   * @param vmId - id of the vm
   * @return Tracking {@link Task}
   * @throws IOException
   */
  public Task performRestartOperation(String vmId) throws IOException {
    String path = String.format("%s/%s/restart", getBasePath(), vmId);

    HttpResponse httpResponse = this.restClient.perform(
        RestClient.Method.POST,
        path,
        null);
    this.restClient.checkResponse(httpResponse, HttpStatus.SC_CREATED);

    return parseTaskFromHttpResponse(httpResponse);
  }

  /**
   * Perform a VM Restart Operation on specified vm.
   * @param vmId
   * @param responseCallback
   * @throws IOException
   */
  public void performRestartOperationAsync(final String vmId,
                                         final FutureCallback<Task> responseCallback)
      throws IOException {
    String path = String.format("%s/%s/restart", getBasePath(), vmId);

    createObjectAsync(path, null, responseCallback);
  }

  /**
   * Perform a VM Resume Operation on specified vm.
   * @param vmId - id of the vm
   * @return Tracking {@link Task}
   * @throws IOException
   */
  public Task performResumeOperation(String vmId) throws IOException {
    String path = String.format("%s/%s/resume", getBasePath(), vmId);

    HttpResponse httpResponse = this.restClient.perform(
        RestClient.Method.POST,
        path,
        null);
    this.restClient.checkResponse(httpResponse, HttpStatus.SC_CREATED);

    return parseTaskFromHttpResponse(httpResponse);
  }

  /**
   * Perform a VM Resume Operation on specified vm.
   * @param vmId
   * @param responseCallback
   * @throws IOException
   */
  public void performResumeOperationAsync(final String vmId,
                                         final FutureCallback<Task> responseCallback)
      throws IOException {
    String path = String.format("%s/%s/resume", getBasePath(), vmId);

    createObjectAsync(path, null, responseCallback);
  }

  /**
   * Perform a VM Suspend Operation on specified vm.
   * @param vmId - id of the vm
   * @return Tracking {@link Task}
   * @throws IOException
   */
  public Task performSuspendOperation(String vmId) throws IOException {
    String path = String.format("%s/%s/suspend", getBasePath(), vmId);

    HttpResponse httpResponse = this.restClient.perform(
        RestClient.Method.POST,
        path,
        null);
    this.restClient.checkResponse(httpResponse, HttpStatus.SC_CREATED);

    return parseTaskFromHttpResponse(httpResponse);
  }

  /**
   * Perform a VM Suspend Operation on specified vm.
   * @param vmId
   * @param responseCallback
   * @throws IOException
   */
  public void performSuspendOperationAsync(final String vmId,
                                         final FutureCallback<Task> responseCallback)
      throws IOException {
    String path = String.format("%s/%s/suspend", getBasePath(), vmId);

    createObjectAsync(path, null, responseCallback);
  }


  /**
   * Attaches a persistent disk to specified VM.
   * @param vmId - id of the vm
   * @param vmDiskOperation {@link VmDiskOperation} specifiying the disk to attach
   * @return Tracking {@link Task}
   * @throws IOException
   */
  public Task attachDisk(String vmId, VmDiskOperation vmDiskOperation) throws IOException {
    String path = String.format("%s/%s/attach_disk", getBasePath(), vmId);

    HttpResponse httpResponse = this.restClient.perform(
        RestClient.Method.POST,
        path,
        serializeObjectAsJson(vmDiskOperation));
    this.restClient.checkResponse(httpResponse, HttpStatus.SC_CREATED);

    return parseTaskFromHttpResponse(httpResponse);
  }

  /**
   * Attaches a persistent disk to specified VM.
   * @param vmId
   * @param vmDiskOperation
   * @param responseCallback
   * @throws IOException
   */
  public void attachDiskAsync(final String vmId, VmDiskOperation vmDiskOperation,
                                    final FutureCallback<Task> responseCallback)
      throws IOException {
    String path = String.format("%s/%s/attach_disk", getBasePath(), vmId);

    createObjectAsync(path, serializeObjectAsJson(vmDiskOperation), responseCallback);
  }

  /**
   * Detaches a persistent disk to specified VM.
   * @param vmId - id of the vm
   * @param vmDiskOperation {@link VmDiskOperation} specifiying the disk to detach
   * @return Tracking {@link Task}
   * @throws IOException
   */
  public Task detachDisk(String vmId, VmDiskOperation vmDiskOperation) throws IOException {
    String path = String.format("%s/%s/detach_disk", getBasePath(), vmId);

    HttpResponse httpResponse = this.restClient.perform(
        RestClient.Method.POST,
        path,
        serializeObjectAsJson(vmDiskOperation));
    this.restClient.checkResponse(httpResponse, HttpStatus.SC_CREATED);

    return parseTaskFromHttpResponse(httpResponse);
  }

  /**
   * Detached a persistent disk to specified VM.
   * @param vmId
   * @param vmDiskOperation
   * @param responseCallback
   * @throws IOException
   */
  public void detachDiskAsync(final String vmId, VmDiskOperation vmDiskOperation,
                              final FutureCallback<Task> responseCallback)
      throws IOException {
    String path = String.format("%s/%s/detach_disk", getBasePath(), vmId);

    createObjectAsync(path, serializeObjectAsJson(vmDiskOperation), responseCallback);
  }

  /**
   * Uploads the iso pointed to by the inputFileName and attaches it to specified vm.
   *
   * @param vmId - id of the vm
   * @param inputFileName - path of the iso to upload and attach
   * @return Tracking {@link Task}
   * @throws IOException
   */
  public Task uploadAndAttachIso(String vmId, String inputFileName) throws IOException {
    String path = String.format("%s/%s/attach_iso", getBasePath(), vmId);

    HttpResponse response = this.restClient.upload(path, inputFileName, ImmutableMap.<String, String>of());

    this.restClient.checkResponse(response, HttpStatus.SC_CREATED);
    return parseTaskFromHttpResponse(response);
  }

  /**
   * Detaches the iso from the specified VM.
   *
   * @param vmId - id of the vm
   * @return Tracking {@link Task}
   * @throws IOException
   */
  public Task detachIso(String vmId) throws IOException {
    String path = String.format("%s/%s/detach_iso", getBasePath(), vmId);
    HttpResponse response = this.restClient.perform(
        RestClient.Method.POST, path, null);

    this.restClient.checkResponse(response, HttpStatus.SC_CREATED);
    return parseTaskFromHttpResponse(response);
  }

  /**
   * Detaches the iso from the specified VM.
   *
   * @param vmId - id of the vm
   * @param responseCallback
   * @throws IOException
   */
  public void detachIsoAsync(final String vmId, final FutureCallback<Task> responseCallback) throws IOException {
    String path = String.format("%s/%s/detach_iso", getBasePath(), vmId);

    createObjectAsync(path, null, responseCallback);
  }

  /**
   * Gets network information for the specified VM.
   *
   * @param vmId - Supplies the ID of the target VM.
   * @return Tracking {@link Task}.
   * @throws IOException
   */
  public Task getNetworks(String vmId) throws IOException {
    String path = String.format("%s/%s/networks", getBasePath(), vmId);

    HttpResponse httpResponse = this.restClient.perform(RestClient.Method.GET, path, null);
    this.restClient.checkResponse(httpResponse, HttpStatus.SC_CREATED);

    return parseTaskFromHttpResponse(httpResponse);
  }

  /**
   * Gets network information for the specified VM.
   *
   * @param vmId - Supplies the ID of the target VM.
   * @param responseCallback
   * @throws IOException
   */
  public void getNetworksAsync(String vmId, FutureCallback<Task> responseCallback) throws IOException {
    String path = String.format("%s/%s/networks", getBasePath(), vmId);

    getObjectByPathAsync(path, HttpStatus.SC_CREATED, responseCallback, new TypeReference<Task>() { });
  }

  public static VmNetworks parseVmNetworksFromTask(Task task) throws IOException {
    // task.getResourceProperties() is a HashSet, we need to serialize it to Json
    // then deserialize with correct type.
    String json = objectMapper.writeValueAsString(task.getResourceProperties());
    return deserializeObjectFromJson(json, new TypeReference<VmNetworks>(){});
  }

  /**
   * Add metadata to the specified VM.
   * @param vmId
   * @param metadata
   * @return Tracking {@link Task}
   * @throws IOException
   */
  public Task setMetadata(final String vmId, final VmMetadata metadata)
      throws IOException {
    String path = String.format("%s/%s/set_metadata", getBasePath(), vmId);

    HttpResponse httpResponse = this.restClient.perform(
        RestClient.Method.POST,
        path,
        serializeObjectAsJson(metadata));
    this.restClient.checkResponse(httpResponse, HttpStatus.SC_CREATED);

    return parseTaskFromHttpResponse(httpResponse);
  }

  /**
   * Add metadata to the specified VM.
   * @param vmId
   * @param metadata
   * @param responseCallback
   * @throws IOException
   */
  public void setMetadataAsync(
      final String vmId,
      final VmMetadata metadata,
      final FutureCallback<Task> responseCallback)
      throws IOException {
    String path = String.format("%s/%s/set_metadata", getBasePath(), vmId);

    createObjectAsync(path, serializeObjectAsJson(metadata), responseCallback);
  }
}

File: tests/common/lib/common/api_client/vm_api.rb
# Copyright 2015 VMware, Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy of
# the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed
# under the License is distributed on an "AS IS" BASIS, without warranties or
# conditions of any kind, EITHER EXPRESS OR IMPLIED. See the License for the
# specific language governing permissions and limitations under the License.

module EsxCloud
  class ApiClient
    module VmApi

      VMS_ROOT = "/vms"

      # @param [String] project_id
      # @param [Hash] payload
      # @return [Vm]
      def create_vm(project_id, payload)
        response = @http_client.post_json("/projects/#{project_id}/vms", payload)
        check_response("Create VM (#{project_id}) #{payload}", response, 201)

        task = poll_response(response)
        find_vm_by_id(task.entity_id)
      end

      # @param [String] id
      # @return [Boolean]
      def delete_vm(id)
        response = @http_client.delete("#{VMS_ROOT}/#{id}")
        check_response("Delete VM '#{id}'", response, 201)

        poll_response(response)
        true
      end

      # @param [String] project_id
      # @return [VmList]
      def find_all_vms(project_id)
        response = @http_client.get("/projects/#{project_id}/vms")
        check_response("Find all VMs for project '#{project_id}'", response, 200)

        VmList.create_from_json(response.body)
      end

      # @param [String] id
      # @return [Vm]
      def find_vm_by_id(id)
        response = @http_client.get("#{VMS_ROOT}/#{id}")
        check_response("Find VM by ID '#{id}'", response, 200)

        Vm.create_from_json(response.body)
      end

      # @param [String] project_id
      # @param [String] name
      # @return [VmList]
      def find_vms_by_name(project_id, name)
        response = @http_client.get("/projects/#{project_id}/vms?name=#{name}")
        check_response("Find VMs by name '#{name}' in project '#{project_id}'", response, 200)

        VmList.create_from_json(response.body)
      end

      # @param [String] id
      # @param [String] state
      # @return [TaskList]
      def get_vm_tasks(id, state = nil)
        url = "#{VMS_ROOT}/#{id}/tasks"
        url += "?state=#{state}" if state
        response = @http_client.get(url)
        check_response("Get tasks for VM '#{id}'", response, 200)

        TaskList.create_from_json(response.body)
      end

      # @param [String] id
      # @return [VmNetworks]
      def get_vm_networks(id)
        url = "#{VMS_ROOT}/#{id}/networks"
        response = @http_client.get(url)
        check_response("Get network connections for VM '#{id}'", response, 201)

        task = poll_response(response)
        VmNetworks.create_from_task(task)
      end

      # @param [String] id
      # @return [MksTicket]
      def get_vm_mks_ticket(id)
        url = "#{VMS_ROOT}/#{id}/mks_ticket"
        response = @http_client.get(url)
        check_response("Get mks ticket for VM '#{id}'", response, 201)

        task = poll_response(response)
        MksTicket.create_from_hash(task.resource_properties)
      end

      # @param [String] id
      # @return [Vm]
      def start_vm(id)
        response = @http_client.post_json("#{VMS_ROOT}/#{id}/start", {})
        check_response("Perform start operation on VM '#{id}'", response, 201)
        task = poll_response(response)
        find_vm_by_id(task.entity_id)
      end

      # @param [String] id
      # @return [Vm]
      def stop_vm(id)
        response = @http_client.post_json("#{VMS_ROOT}/#{id}/stop", {})
        check_response("Perform stop operation on VM '#{id}'", response, 201)
        task = poll_response(response)
        find_vm_by_id(task.entity_id)
      end

      # @param [String] id
      # @return [Vm]
      def restart_vm(id)
        response = @http_client.post_json("#{VMS_ROOT}/#{id}/restart", {})
        check_response("Perform restart operation on VM '#{id}'", response, 201)
        task = poll_response(response)
        find_vm_by_id(task.entity_id)
      end

      # @param [String] id
      # @return [Vm]
      def resume_vm(id)
        response = @http_client.post_json("#{VMS_ROOT}/#{id}/resume", {})
        check_response("Perform resume operation on VM '#{id}'", response, 201)
        task = poll_response(response)
        find_vm_by_id(task.entity_id)
      end

      # @param [String] id
      # @return [Vm]
      def suspend_vm(id)
        response = @http_client.post_json("#{VMS_ROOT}/#{id}/suspend", {})
        check_response("Perform suspend operation on VM '#{id}'", response, 201)
        task = poll_response(response)
        find_vm_by_id(task.entity_id)
      end

      # @param [String] id
      # @param [String] operation
      # @param [Hash] args
      # @return [Vm]
      def perform_vm_disk_operation(id, operation, disk_id, args = {})
        op = operation.to_s.downcase
        payload = {diskId: disk_id, arguments: args}
        response = @http_client.post_json("#{VMS_ROOT}/#{id}/#{op}", payload)
        check_response("Perform VM disk operation #{op} #{payload} on VM '#{id}'", response, 201)

        task = poll_response(response)
        find_vm_by_id(task.entity_id)
      end

      # @param [String] id
      # @param [String] iso_path
      # @param [String] iso_name
      # @return [Vm]
      def perform_vm_iso_attach(id, iso_path, iso_name = nil)
        response = @http_client.upload("#{VMS_ROOT}/#{id}/attach_iso", iso_path, iso_name)
        check_response("Attach ISO '#{iso_path}' to VM", response, 201)

        task = poll_response(response)
        find_vm_by_id(task.entity_id)
      end

      # @param [String] id
      # @return [Vm]
      def perform_vm_iso_detach(id)
        response = @http_client.post_json("#{VMS_ROOT}/#{id}/detach_iso", nil)
        check_response("Detach ISO to VM #{id}", response, 201)

        task = poll_response(response)
        find_vm_by_id(task.entity_id)
      end

      # @param [String] id
      # @param [Hash] payload
      # @return [Vm]
      def perform_vm_metadata_set(id, payload)
        response = @http_client.post_json("#{VMS_ROOT}/#{id}/set_metadata", payload)
        check_response("Set Metadata to VM #{id}", response, 201)

        task = poll_response(response)
        find_vm_by_id(task.entity_id)
      end

      # @param [String] id
      # @param [Hash] payload
      # @return [Vm]
      def acquire_floating_ip(id, payload)
        response = @http_client.post_json("#{VMS_ROOT}/#{id}/acquire_floating_ip", payload)
        check_response("Acquire floating IP for VM #{id}", response, 201)

        task = poll_response(response)
        find_vm_by_id(task.entity_id)
      end

      # @param [Stirng] id
      # @return [Vm]
      def release_floating_ip(id)
        response = @http_client.delete("#{VMS_ROOT}/#{id}/release_floating_ip")
        check_response("Release floating IP from VM #{id}", response, 201)

        task = poll_response(response)
        find_vm_by_id(task.entity_id)
      end
    end
  end
end
